### PR TITLE
Add verify()

### PR DESCRIPTION
**Pull request by JSAssassin**
addresses #5

### CODE REVIEW

File: lib/httpDigest.js [hunk 1]
```diff
@@ -19,13 +22,54 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+    data = JSON.stringify(data);
+  }
   let digest;
+  let encoded;
+  const algorithm = header.split('=')[0].replace('-', '').toLowerCase();
+  if(algorithm === 'sha256') {
+    encoded = header.substring(header.indexOf('=') + 1);
+    digest = await _getDigest({data, algorithm});
+  } else if(algorithm === 'mh') {
+    encoded = header.substring(header.indexOf('=') + 1).slice(1);
```
Comment (dlongley) [inline lib/httpDigest.js:54]: How confident are we in these parsing rules? Is there a tiny http parsing lib we can rely on instead?

File: lib/httpDigest.js [hunk 2]
```diff
@@ -19,13 +22,54 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+    digest = await _createMultiHash({digest});
+  }
+  const hashedBytes = new Uint8Array(base64url.decode(encoded));
```
Comment (dlongley) [inline lib/httpDigest.js:57]: Is this right for multihash or is it multibase encoded?

File: lib/httpDigest.js [hunk 3]
```diff
@@ -19,13 +22,54 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+
+  if(JSON.stringify(digest) !== JSON.stringify(hashedBytes)) {
```
Comment (dlongley) [inline lib/httpDigest.js:59]: Why are we stringifying here instead of comparing bytes? If we're going to compare as strings, we could just encode the digest we produced and compare against the given value, which means we don't have to decode at all, right? Note: I do think a non-timing comparison here is fine because none of the values are secret.

File: lib/httpDigest.js (hunk 1 above)
Comment (dlongley) [reply lib/httpDigest.js:54]: Also, this isn't properly checking the multibase char for `u` as an indication of base64url encoding.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/httpDigest.js (hunk 3 above)
Comment (dmitrizagidulin) [reply lib/httpDigest.js:59]: I suspect it's because we've had trouble with cross-platform buffer/uint8array comparison without strings. (We ended up using stringify in our crypto libs as well, like https://github.com/digitalbazaar/ed25519-verification-key-2018/blob/master/src/Ed25519VerificationKey2018.js#L315 )

File: lib/httpDigest.js (hunk 1 above)
Comment (dmitrizagidulin) [reply lib/httpDigest.js:54]: > Also, this isn't properly checking the multibase char for u as an indication of base64url encoding.

+1, should test for that.

Comment (dmitrizagidulin) [reply lib/httpDigest.js:54]: > How confident are we in these parsing rules? Is there a tiny http parsing lib we can rely on instead?

Not super. (Obv we can get more confident with more tests.) But yeah, if there was a generic http header parsing lib (for that format specifically), might be better. (Not sure if there is one)

File: lib/httpDigest.js (hunk 3 above)
Comment (JSAssassin) [reply lib/httpDigest.js:59]: @dlongley @dmitrizagidulin yeah I wasn't able figure a way to compare buffer with unit8array, so I converted both to unit8array and compared it.

Comment (dlongley) [reply lib/httpDigest.js:59]: @dmitrizagidulin, please file an issue to fix that verification key thing.

@JSAssassin, let's just encode the digest we get and then compare the encoded strings -- then we can skip decoding.

File: lib/httpDigest.js (hunk 1 above)
Comment (dlongley) [reply lib/httpDigest.js:54]: If we're not fixing this now, let's make sure we've got good tests and add an issue for looking into a better/more confident parser lib/rules.

File: lib/httpDigest.js (hunk 3 above)
Comment (JSAssassin) [reply lib/httpDigest.js]: @dlongley Updated - 5647cd2

File: lib/base64url.js [hunk 4]
```diff
@@ -1,5 +1,5 @@
 /*!
  * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
  */
-import {encode} from 'base64url-universal';
-export default {encode};
+import {encode, decode} from 'base64url-universal';
+export default {encode, decode};
```
Comment (dlongley) [inline lib/base64url.js:5]: Technically don't need `decode` anymore. Also, we don't need this extra file, just import directly.

File: lib/httpDigest.js [hunk 5]
```diff
@@ -19,13 +22,46 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+ * @param {object} options - The options to use.
+ * @param {string|object} options.data - The data to be verified.
+ * @param {string} options.header - The digest header string to
+ *   be verified against.
+ *
+ * @returns {object} - Returns an object with verified true or false.
+ */
+async function verify({data, header}) {
   let digest;
+  if(header.startsWith('SHA-256')) {
```
Comment (dlongley) [inline lib/httpDigest.js:45]: Header keys are case insensitive. We need to parse the key and value here, convert the key to lowercase as before and compare. Then, we'll want to compare the generated, encoded, digest value to the parsed value -- not the entire header. I get that there's an effort to reuse as much code as we can here, but some tweaks are still needed or this (and the comparison below) won't function properly. +1 for figuring out how to do that in the most DRY way.

**Review by dlongley — CHANGES_REQUESTED** [review]

File: lib/httpDigest.js (hunk 5 above)
Comment (dlongley) [reply lib/httpDigest.js:45]: Possible suggestion would be to have a helper that returns the digest as a key + value pairing ... and just have `create` join that with `=` ... and call the helper within `verify` without joining, then you can compare the values only.

File: lib/httpDigest.js (hunk 3 above)
Comment (dmitrizagidulin) [reply lib/httpDigest.js]: @dlongley 

> @dmitrizagidulin, please file an issue to fix that verification key thing.

Can you say more? What's the part that should be fixed?

Comment (dlongley) [reply lib/httpDigest.js]: Filed an issue here: https://github.com/digitalbazaar/ed25519-verification-key-2018/issues/9

File: lib/httpDigest.js [hunk 6]
```diff
@@ -19,13 +22,64 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+
+/**
+ * Verifies digest of data created by create().
+ *
+ * @param {object} options - The options to use.
+ * @param {string|object} options.data - The data to be verified.
+ * @param {string} options.header - The digest header string to
+ *   be verified against.
+ *
+ * @returns {object} - Returns an object with verified true or false.
```
Comment (dlongley) [inline lib/httpDigest.js:41]:
```suggestion
 * @returns {object} - An object with `verified` set to `true` or `false`.
```

File: lib/httpDigest.js [hunk 7]
```diff
@@ -19,13 +22,64 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+ */
+async function verify({data, header}) {
```
Comment (dlongley) [inline lib/httpDigest.js:43]: Similar fix needed above on `create`:

```suggestion
async function verify({data, header} = {}) {
```

File: lib/httpDigest.js [hunk 8]
```diff
@@ -19,13 +22,64 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+    const {value} = _parseHashedString(hashedDigestString);
+    hashedDigestStringValue = value;
+  } else {
+    throw new Error(`${algorithm} algorithm is not unsupported.`);
+  }
+
+  return {verified: headerValue === hashedDigestStringValue};
+}
+
+function _parseHashedString(hashedString) {
```
Comment (dlongley) [inline lib/httpDigest.js:64]: I think this would be a better name:

```suggestion
function _parseHeader(hashedString) {
```

File: lib/httpDigest.js [hunk 9]
```diff
@@ -19,13 +22,64 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+ * @returns {object} - Returns an object with verified true or false.
+ */
+async function verify({data, header}) {
+  const {algorithm, value: headerValue} = _parseHashedString(header);
+  let hashedDigestString;
+  let hashedDigestStringValue;
+
+  if(algorithm == 'sha256') {
+    hashedDigestString = await create({
+      data, algorithm, useMultihash: false});
```
Comment (dlongley) [inline lib/httpDigest.js:50]:
```suggestion
    hashedDigestString = await create(
      {data, algorithm, useMultihash: false});
```

File: lib/httpDigest.js [hunk 10]
```diff
@@ -19,13 +22,64 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+ *   be verified against.
+ *
+ * @returns {object} - Returns an object with verified true or false.
+ */
+async function verify({data, header}) {
+  const {algorithm, value: headerValue} = _parseHashedString(header);
+  let hashedDigestString;
+  let hashedDigestStringValue;
+
+  if(algorithm == 'sha256') {
```
Comment (dlongley) [inline lib/httpDigest.js:48]:
```suggestion
  if(algorithm === 'sha256') {
```

File: lib/httpDigest.js [hunk 11]
```diff
@@ -19,13 +22,64 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+    hashedDigestString = await create({
+      data, algorithm, useMultihash: false});
+    const {value} = _parseHashedString(hashedDigestString);
+    hashedDigestStringValue = value;
+  } else if(algorithm == 'mh') {
```
Comment (dlongley) [inline lib/httpDigest.js:53]:
```suggestion
  } else if(algorithm === 'mh') {
```

File: lib/httpDigest.js [hunk 12]
```diff
@@ -35,21 +89,21 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
-    mh[0] = 0x12;
-    mh[1] = 0x20;
-    mh.set(digest, 2);
-    // encode multihash using multibase, base64url: `u`
-    return `mh=u${base64url.encode(mh)}`;
-  }
-
-  return `SHA-256=${base64.encode(digest)}`;
-};
+function _createMultiHash({digest}) {
```
Comment (dlongley) [inline lib/httpDigest.js:95]:
```suggestion
function _createMultihash({digest}) {
```

File: lib/httpDigest.js [hunk 13]
```diff
@@ -19,13 +22,64 @@ async function create({data, algorithm = 'sha256', useMultihash = true}) {
...
+ * @param {string|object} options.data - The data to be verified.
+ * @param {string} options.header - The digest header string to
+ *   be verified against.
+ *
+ * @returns {object} - Returns an object with verified true or false.
+ */
+async function verify({data, header}) {
+  const {algorithm, value: headerValue} = _parseHashedString(header);
+  let hashedDigestString;
+  let hashedDigestStringValue;
```
Comment (dlongley) [inline lib/httpDigest.js:46]: These variable names are hard to follow... and I think they are probably the result of not breaking this up into its logical pieces quite right. I think this is too hard to make suggestions for, so I'll push a commit to take a look at.

**Review by dlongley — CHANGES_REQUESTED** [review]
I'll work on a commit here.

**Review by dlongley — APPROVED** [review]
@JSAssassin, @dmitrizagidulin -- I pushed a couple of commits with the naming and restructuring I was looking for. The changes focus on making the names/functions work with the logical components more precisely (i.e., this library is working with "header values" for the "Digest" header, and those values are made up of a key -- that either refers to the algorithm directly or refers to a multihash whereby the algorithm is in the digest itself -- and an encoded digest).

This approach also includes a helper for creating these components so we don't have to create a full header value and then parse it all over again just to compare encoded digests.

**Review by dmitrizagidulin — APPROVED** [review]
👍

File: lib/httpDigest.js [hunk 14]
```diff
@@ -1,55 +1,100 @@
...
+
+/**
+ * Verifies the HTTP `Digest` header value against the given HTTP body `data`.
+ *
+ * @param {object} options - The options to use.
+ * @param {string|object} options.data - The data to be verified.
+ * @param {string} options.headerValue - The digest header value to verify
+ *   the data against.
+ *
+ * @returns {object} - An object with `verified` set to `true` or `false`.
```
Comment (mattcollier) [inline lib/httpDigest.js:36]: Needs Promise annotation as with other async APIs here.

File: lib/httpDigest.js [hunk 15]
```diff
@@ -1,55 +1,100 @@
...
+ */
+export async function verifyHeaderValue({data, headerValue} = {}) {
+  const {key, algorithm, encodedDigest} = _parseHeaderValue(headerValue);
+  const {encodedDigest: expectedDigest} = await _createHeaderValueComponents(
+    {data, algorithm, useMultihash: key === 'mh'});
+  return {verified: encodedDigest === expectedDigest};
+}
```
Comment (mattcollier) [inline lib/httpDigest.js:43]: @dlongley do you agree with the following?

We have adopted the API contract for `verifyFoo` APIs that they do not throw.  They always return a true/false value along with error report as appropriate.

I recommend that all helpers being called here be wrapped in a try catch and...

The verify pattern used elsewhere is that all extraneous error will end up producing a return value something like:
```
{
  verified: false,
  error: e
}
```

File: lib/httpDigest.js [hunk 16]
```diff
@@ -1,55 +1,100 @@
...
-  const encodedData = new TextEncoder().encode(data);
-  let digest;
+  if(algorithm !== 'sha256') {
+    throw new Error(`Algorithm "${algorithm}" is not supported.`);
+  }
+  const digest = await _getDigest({data, algorithm});
+  if(useMultihash) {
+    return {key: 'mh', encodedDigest: _createMultihash({digest})};
+  }
+  return {key: 'SHA-256', encodedDigest: base64Encode(digest)};
```
Comment (mattcollier) [inline lib/httpDigest.js:57]: I feel like the `key` value being returned here should be informed by the return value of `_getDigest` on L53.  I would like to see an easier transition when the time comes to add additional methods.  Possibly even including string literal `sha256` being in some CONST array of accepted digest methods.

looking further down, i see a possibility that the data structure for a hash algorithm might also include the expected multibase prefix instead of using string literals in code below in `_parseHeaderValue`.

File: lib/httpDigest.js [hunk 17]
```diff
@@ -1,55 +1,100 @@
...
-  switch(algorithm) {
-    case 'sha256': {
-      digest = new Uint8Array(
-        await crypto.subtle.digest({name: 'SHA-256'}, encodedData));
-      break;
-    }
-    case 'hs2019': {
-      // todo: support hs2019 / SHA-512
+function _parseHeaderValue(headerValue) {
+  const [key, ...rest] = headerValue.split('=');
```
Comment (mattcollier) [inline lib/httpDigest.js:61]: I'm trying to imagine what the data format coming in here is.  Can you provide a sample?

File: lib/httpDigest.js [hunk 18]
```diff
@@ -1,55 +1,100 @@
...
+  throw new Error(`Algorithm "${algorithm}" is not unsupported.`);
+}
 
-  return `SHA-256=${base64.encode(digest)}`;
-};
-
-export default {
-  create
-};
+function _createMultihash({digest}) {
```
Comment (mattcollier) [inline lib/httpDigest.js:91]: would there be any benefit to using the npm packages for this and other multihash operations?

File: lib/index.js [hunk 19]
```diff
@@ -1,6 +1,4 @@
 /*!
```
Comment (mattcollier) [inline lib/index.js:1]: confused by what's going on here with index.js and main.js, the pattern is the reverse of what is normally done.  By convention I believe the commonjs code is index.js and the browser/ES6 implementation is main.js.

**Review by mattcollier — CHANGES_REQUESTED** [review]

File: lib/httpDigest.js (hunk 17 above)
Comment (JSAssassin) [reply lib/httpDigest.js:61]: @mattcollier Here's what the data format looks like.


```
key: SHA-256
rest: X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE
```

File: lib/httpDigest.js (hunk 14 above)
Comment (JSAssassin) [reply lib/httpDigest.js]: Updated - dad4dbf

File: lib/index.js (hunk 19 above)
Comment (JSAssassin) [reply lib/index.js:1]: Updated - dad4dbf

File: lib/httpDigest.js (hunk 18 above)
Comment (JSAssassin) [reply lib/httpDigest.js:92]: @dlongley @dmitrizagidulin ^

Comment (dmitrizagidulin) [reply lib/httpDigest.js:92]: @mattcollier - it's a tough call, I always struggle with that. On the one hand, we should support all sorts of multihash algs, that's what multihash is _for_, right? But on the other, I'm always reluctant to pay the extra library size, especially for isomorphic libs.

@dlongley ?

File: lib/httpDigest.js (hunk 17 above)
Comment (mattcollier) [reply lib/httpDigest.js:62]: @JSAssassin thank you.  I made a replit to try and make sense of this parsing algorithm: https://repl.it/@mattcollier/FavoriteUnfoldedDemand#index.js

is the output `X48E9qOokqqrvdts8nOJRJN3OWDUoyWxBf7kbu9DBPE=` (with a trailing equal) the expected value of `encodedDigest` on L63?

File: lib/httpDigest.js (hunk 18 above)
Comment (dlongley) [reply lib/httpDigest.js:92]: Don't want to pay for cost of the lib -- also don't know if the libs are isomorphic.

File: lib/httpDigest.js (hunk 17 above)
Comment (JSAssassin) [reply lib/httpDigest.js:62]: @mattcollier from my understanding, that's correct. The expected value of encodedDigest should have a trailing equal.

Note: updated how rest looks in my previous comment(made an error earlier)

Comment (mattcollier) [reply lib/httpDigest.js:62]: In that case, I would like to offer this alternative implementation that I think is easier to reason about: https://repl.it/@mattcollier/alternateSplit

Comment (JSAssassin) [reply lib/httpDigest.js]: Updated - 4db2f55

File: lib/httpDigest.js (hunk 15 above)
Comment (JSAssassin) [reply lib/httpDigest.js]: @mattcollier @dlongley I checked and it looks like that pattern is being used in other verifyFoos in other repos( for example [here](https://github.com/digitalbazaar/jsonld-signatures/blob/d81e605b910d5f5bcce52edac2334ff7b89212df/lib/suites/LinkedDataSignature2015.js#L80)). So I updated it - dad4dbf.

**Review by mattcollier — APPROVED** [review]

**Comment by mattcollier** [comment]
@JSAssassin this needs a changelog entry.

File: .github/workflows/main.yml [hunk 20]
```diff
@@ -0,0 +1,74 @@
...
+    steps:
+    - uses: actions/checkout@v2
+    - name: Use Node.js ${{ matrix.node-version }}
+      uses: actions/setup-node@v1
+      with:
+        node-version: ${{ matrix.node-version }}
+    - name: Start MongoDB v${{ matrix.mongodb-version}}
+      uses: supercharge/mongodb-github-action@1.3.0
+      with:
+        mongodb-version: ${{ matrix.mongodb-version }}
```
Comment (mattcollier) [inline .github/workflows/main.yml:22]: Is this needed?  I don't think this test suite interacts with mongo.

File: .github/workflows/main.yml [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+      run: npm run lint
+  coverage:
+    needs: [test-node]
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    services:
+      mongodb:
+        image: mongo:4.2
+        ports:
+          - 27017:27017
```
Comment (mattcollier) [inline .github/workflows/main.yml:53]: again, no need for mongo?

File: .github/workflows/main.yml [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+      uses: actions/setup-node@v1
+      with:
+        node-version: ${{ matrix.node-version }}
+    - run: |
+        npm install
+    - name: Generate coverage report
+      run: |
+        npm run coverage-ci
+      env:
+        CI: true
```
Comment (mattcollier) [inline .github/workflows/main.yml:69]: remove.

**Review by mattcollier — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+on: [push]
+
+jobs:
+  test-node:
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+    strategy:
+      matrix:
+        node-version: [12.x, 14.x]
+        mongodb-version: ['4.2']
```
Comment (mattcollier) [inline .github/workflows/main.yml:12]: not needed.  I will add... that I don't know where the matrix mongodb pattern came from, but I don't think we have a use case for that at this time.

**Review by mattcollier — APPROVED** [review]
One more minor change to workflow.  TY
